src/ports/ORMPUpgradeablePort.sol

```diff
@@ -0,0 +1,125 @@
+// This file is part of Darwinia.
+// Copyright (C) 2018-2023 Darwinia Network
+// SPDX-License-Identifier: GPL-3.0
+//
+// Darwinia is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Darwinia is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Darwinia. If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity ^0.8.17;
+
+import "./base/BaseMessagePort.sol";
+import "./base/PortLookup.sol";
+import "ORMP/src/interfaces/IORMP.sol";
+import "ORMP/src/user/AppBase.sol";
+import "@openzeppelin/contracts/access/Ownable2Step.sol";
+import "@openzeppelin/contracts/utils/structs/EnumerableSet.sol";
+
+contract ORMPUpgradeablePort is Ownable2Step, AppBase, BaseMessagePort, PortLookup {
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    address public ormp;
+
+    mapping(bytes32 => bool) public dones;
+
+    EnumerableSet.AddressSet internal historyORMPSet;
+
+    event SetORMP(address previousORMP, address currentORMP);
+
+    modifier onlyORMP() override {
+        require(historyORMPSet.contains(msg.sender), "!ormps");
+        _;
+    }
+
+    constructor(address dao, address ormp_, string memory name) BaseMessagePort(name) {
+        _transferOwnership(dao);
+        ormp = ormp_;
+        historyORMPSet.add(ormp_);
+    }
+
+    /// @notice How to migrate to new ORMP contract.
+    /// 1. setORMP to new ORMP contract.
+    /// 2. delete previousORMP after relay on-flight message.
+    function setORMP(address ormp_) external onlyOwner {
```

Comment:
1. Deploy ORMP, Oracle and Relayer contract.
   - ORMPipe config: oracle config
3. setORMP new ORMP contract.
4. disable previousORMP send interface.
5. relay all on-flight message.
6. delete previouseORMP
7. disable default relayer and oracle.